backends/ebpf/psa/README.md

```diff
@@ -0,0 +1,233 @@
+
+### psabpf API and psabpf-ctl
+
+We provide the `psabpf` C API and the `psabpf-ctl` CLI tool that can be used to manage eBPF programs generated by P4-eBPF compiler.
+To install the CLI tool, follow the guide in [the psabpf repository](https://github.com/P4-Research/psabpf). Use `psabpf-ctl help` to get all possible commands.
+
+**Note!** Although eBPF objects can be loaded and managed by other tools (e.g. `bpftool`), we recommend using `psabpf-ctl`. Some features
+(e.g., default actions) will only work when using `psabpf-ctl`.
+
+To load eBPF programs generated by P4-eBPF compiler run:
+
+```bash
+psabpf-ctl pipeline load id <ID> out.o
+```
+
+Then, for each interface that should be attached to PSA-eBPF run:
+
+```bash
+psabpf-ctl add-port pipe <ID> dev <INTF>
+```
+
+> Note! Make sure that the BPF filesystem is mounted under `/sys/fs/bpf`.
+
+## Troubleshooting
+
+The PSA implementation for eBPF backend generates standard BPF objects that can be inspected using `bpftool`.
+
+To troubleshoot PSA-eBPF program you will probably need `bpftool`. Follow the steps below to install it.
+
+You should be able to see `bpftool help`:
+
+```bash
+$ bpftool help
+  Usage: bpftool [OPTIONS] OBJECT { COMMAND | help }
+         bpftool batch file FILE
+         bpftool version
+  
+         OBJECT := { prog | map | link | cgroup | perf | net | feature | btf | gen | struct_ops | iter }
+         OPTIONS := { {-j|--json} [{-p|--pretty}] | {-f|--bpffs} |
+                      {-m|--mapcompat} | {-n|--nomount} }
+```
+
+Refer to [the bpftool guide](https://manpages.ubuntu.com/manpages/focal/man8/bpftool-prog.8.html) for more examples how to use it.
+
+# TODO / Limitations
+
+We list the known bugs/limitations below. Refer to the Roadmap section for features planned in the near future.
+
+- Larger bit fields (e.g. IPv6 addresses) may not work properly
+- We noticed that `bpf_xdp_adjust_meta()` isn't implemented by some NIC drivers, so the PSA impl may not work 
+with some NICs. So far, we have verified the correct behavior with Intel 82599ES.
+- `lookahead()` with bit fields (e.g., `bit<16>`) doesn't work.
+- `@atomic` operation is not supported yet.
+- `psa_idle_timeout` is not supported yet. 
+
+# Roadmap
+
+## Planned features
+
+All the below features are already implemented and will be contributed to the P4 compiler in subsequent pull requests.
+
+- **XDP support.** The current version of P4-eBPF compiler leverages the BPF TC hook for P4-programmable packet processing. 
+The TC subsystem enables implementation of the full PSA specification, contrary to XDP, but offers lower throughput. We're going to 
+contribute the XDP-based version of P4-eBPF that is not fully spec-compliant, but provides higher throughput.
+- **XDP2TC mode.** the generated eBPF programs use `bpf_xdp_adjust_meta()` to transfer original EtherType from XDP to TC. This mode 
+may not be supported by some NIC drivers. We will add two other modes that can be used alternatively. 
```

Comment:
Well.. auto-detection might be tricky, since the P4 compiler can be run on a different machine than the one where the eBPF program is actually deployed. The XDP2TC mode will be configured with compiler flags.